src/dmd/root/aav.d

```diff
@@ -170,3 +170,42 @@ unittest
     v = dmd_aaGetRvalue(aa, null);
     assert(v == cast(void*)3);
 }
+
+struct AssocArray(K,V)
+{
+    private AA* aa;
+
+    /**
+    Returns: The number of key/value pairs.
+    */
+    @property size_t length() const { return dmd_aaLen(aa); }
+
+    /**
+    Lookup value associated with `key` and return an address to it. If the `key`
+    has not been added, it adds it and returns an address to the new value.
+
+    Params:
+        key = key to lookup the value for
+
+    Returns: an address to the value associated with `key`. If `key` does not exists, it
+             is added and an address to the new value is returned.
+    */
+    V* getLvalue(const(K) key)
+    {
+        return cast(V*)dmd_aaGet(&aa, cast(void*)key);
+    }
+
+    /**
+    Lookup and return the value associated with `key`, if the `key` has not been
+    added, it returns null.
+
+    Params:
+        key = key to lookup the value for
+
+    Returns: the value associated with `key` if present, otherwise, null.
+    */
+    V opIndex(const(K) key)
+    {
+        return cast(V)dmd_aaGetRvalue(aa, cast(void*)key);
+    }
```

Comment:
How about adding `opIndexAssign` too?

```d
    auto opIndexAssign(V value, K key)
    {
        auto ptr = dmd_aaGet(&aa, cast(void*) key);
        *ptr = cast(void*) value;
        return *ptr;
    }
```

Reply:
Currently no code uses it this way and I don't anticipate code ever doing this.
```
aa[key] = value;
```
This being the more "intuitive" usage, I have to assume that Walter deliberately chose not to support this.

Reply:
maybe it's just because it wasn't implemented? maybe ask @WalterBright ?

Reply:
I don't think it's something worth bugging him about.  Currently there's no use for it.  We can always add it later if a use comes to light.

Reply:
Keep in mind that the current code base was translated from C+ (not a typo).  I don't think there was any deliberate effort by Walter to avoid it.

Reply:
I'm also basing this on comments he made about one of @ibuclaw's PRs where he made a change that ended up causing multiple hashes of the same key.  The current implementation uses `getLvalue` which allows the developer to do multiple operations on a value with only one lookup.  I think the API was designed to encourage this behavior, and adding `opIndexAssign` would provide a way to circumvent it, making it easy for developers to accidentally write code that has to hash/lookup multiple times.

Reply:
Here's Walter's comment: https://github.com/dlang/dmd/pull/7696#discussion_r161407629